src/operator/contrib/index_copy.cc

```diff
@@ -26,6 +26,99 @@
 namespace mxnet {
 namespace op {
 
+struct index_copy_fwd_cpu {
+  template<typename DType, typename IType>
+  static void Map(int i,
+                  const DType* new_tensor,
+                  const IType* idx,
+                  DType* out_tensor,
+                  int dim_size) {
+    DType* out_ptr = out_tensor + static_cast<int>(idx[i]) * dim_size;
+    const DType* new_ptr = new_tensor + i * dim_size;
+    std::memcpy(out_ptr, new_ptr, sizeof(DType) * dim_size);
+  }
+};
+
+template<>
+void IndexCopyForward<cpu>(const nnvm::NodeAttrs& attrs,
+                           const OpContext& ctx,
+                           const std::vector<TBlob>& inputs,
+                           const std::vector<OpReqType>& req,
+                           const std::vector<TBlob>& outputs) {
+  using namespace mshadow;
+  using namespace mxnet_op;
+  CHECK_EQ(inputs.size(), 3U);
+  CHECK_EQ(outputs.size(), 1U);
+  CHECK_EQ(req.size(), 1U);
+  mshadow::Stream<cpu> *s = ctx.get_stream<cpu>();
+  const TBlob& out = outputs[0];
+  const TBlob& original_tensor = inputs[0];
+  const TBlob& idx_vector = inputs[1];
+  const TBlob& copied_tensor = inputs[2];
+  int dim_size = inputs[2].Size() / inputs[1].Size();
+  // copy original tensor to output
+  copy(s, out, original_tensor);
+  // index copy
+  MSHADOW_TYPE_SWITCH(out.type_flag_, DType, {
+    MSHADOW_TYPE_SWITCH(idx_vector.type_flag_, IType, {
+      Kernel<index_copy_fwd_cpu, cpu>::Launch(
+        s, idx_vector.Size(), copied_tensor.dptr<DType>(),
+        idx_vector.dptr<IType>(), out.dptr<DType>(), dim_size);
+    });
+  });
+}
+
+struct index_copy_bwd_cpu {
+  template<typename DType, typename IType>
+  static void Map(int i,
+                  const DType* out_tensor_grad,
+                  DType* orig_tensor_grad,
+                  DType* new_tensor_grad,
+                  const IType* idx,
+                  int dim_size,
+                  int idx_size) {
+    const DType* out_ptr = out_tensor_grad + i * dim_size;
+    DType* orig_ptr = orig_tensor_grad + i * dim_size;
+    std::memcpy(orig_ptr, out_ptr, sizeof(DType) * dim_size);
+    if (i < idx_size) {
```

Comment:
It puzzled me since the meanings of variable i are different.
I recommend to use copy function to copy out_tensor_grad to orig_tensor_grad, then call the latter code.

Reply:
There is a test case: index[0]=3，i=0,1,2,3
orig_tenser_grad[3] will be rewritten when i==3

Reply:
okay